src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs

```diff
@@ -7,12 +7,25 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using ICSharpCode.Decompiler;
+using ICSharpCode.Decompiler.CSharp;
+using ICSharpCode.Decompiler.TypeSystem;
+using Microsoft.CodeAnalysis.CSharp.Formatting;
+using Microsoft.CodeAnalysis.ErrorReporting;
 using Microsoft.CodeAnalysis.MetadataAsSource;
+using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.SymbolMapping;
 using Microsoft.CodeAnalysis.Text;
 using Microsoft.VisualStudio.Text;
+using Mono.Cecil;
 using Roslyn.Utilities;
+using BraceStyle = ICSharpCode.Decompiler.CSharp.OutputVisitor.BraceStyle;
+using EmptyLineFormatting = ICSharpCode.Decompiler.CSharp.OutputVisitor.EmptyLineFormatting;
+using NewLinePlacement = ICSharpCode.Decompiler.CSharp.OutputVisitor.NewLinePlacement;
+using PropertyFormatting = ICSharpCode.Decompiler.CSharp.OutputVisitor.PropertyFormatting;
+using UsingPlacement = ICSharpCode.Decompiler.CSharp.OutputVisitor.UsingPlacement;
+using Wrapping = ICSharpCode.Decompiler.CSharp.OutputVisitor.Wrapping;
```

Comment:
This is interesting that we're at a core level, but we're referencing a C# decompiler.  Would we not want to make it possible to pull in an arbitrary decompiler so that we could get one matchign the calling language (possibly falling back to C# if we can't find one).  

That way, VB could get VB decompilation, F# could get F# decompilation, etc. etc.

Reply:
➡️ Right now we only have the C# decompiler ready to use. This may change in the future.

Reply:
It would be nice to make this use an ILangaugeService where the default impl of that ILanguageService did hte C# decompilation.  That way our other languages have a natural way to slot in their decompilation without requiring any changes on our side.

Reply:
it would be nice to actually wrap ICSharpCode.Decompiler under ILanguageService rather than exposing it directly in metadata as srouce file service.

Reply:
also, in core layer, exposing something a language specific is violation of layers. things specific for a language (csharp in this case) should be in EditorFeatures.CSharp

Reply:
Note: it's a layering violation to hit a Roslyn.CSharp assembly from a core assembly.  It's not a layering violation (IMO) to hit an arbitrary 3rd party assembly that happens to have C# affinity.  They're just like a simple 3rd party API that converts IL to strings in some form (in this case, that happens to be C#... but that's effectively an implementation detail).  

So i don't mind EditorFeatures referring to Mono/Cecil here.  I also don't mind that we eventually load in Roslyn.CSHarp dlls here (even if we were only referenced by VB) since we'll have to load those dlls anyways to properly display/colorize the file.  

However, that loading should be done through ILanguageServices IMO rather than a direct reference.